### ISSUE TITLE
minimal package installs

### ISSUE THREAD

**Issue by getify** [npm/npm#5264]
I read this [blog post](http://www.rudeshko.com/web/2014/05/13/help-people-consume-your-npm-packages.html) about npm packages having lots of optional bloat in them that puts an an additional disk-space/bandwidth burden on those depending on the package, especially in CI envs where re-builds happen constantly. I myself have experienced this problem before.

I was inspired with an idea similar to the `devDependencies`, but for contents of your package directly (not your dependencies list), which I call "minimal-package". So, I jotted down a quick proposal for adding an extra [minimal-package layer](https://gist.github.com/getify/86683eeaa13ac7474f04) to npm packages.

FTA:

## TL;DR

Add `.npmignore-minimal` file with things in your package that are not-strictly-necessary (readmes, tests, etc) to filter out of minimal-package installs. `npm install` still installs full package. `npm min-install` (or `npm install --production`) installs only the filtered down package contents (aka "minimal-package").

### Bike vs. Shed

As to be expected, there's already been a fair amount of bikeshedding on the syntax (which I don't care that much about) since I tweeted out about this. I hope the syntax/implementation I suggest can be looked past and discussion can be substantively about the utility of the feature/semantics. :)

**Comment by Zarathecat** [comment]
Has there been any further work on this?

**Comment by stefanpenner** [comment]
> I spent considerable amount of time getting the total dep size of grunt/bower/yeoman down last year, by opening a gadzillion PRs with .npmignore. In a project using grunt (+ some grunt plugin) and bower the size of the deps went from ~200MB to ~32MB (4 months later when deps had propagated).

i wanted to thank you @sindresorhus for this effort, I am sure ember-cli is benefiting from your hard work. I have also spent considerable time working to reduce payload sizes, its tedious..

**Comment by getify** [comment]
Bump.

**Comment by coupeez** [comment]
Yeah, bump. 

I'm building apps with nwjs, where a node_modules folder is required for Mac, Linux, Win32 and Win64 builds. The redundancy is huge!

**Comment by tohagan** [comment]
+1 Badly needed.

**Comment by emragins** [comment]
I would really love to see some progress on this.  The past couple times I've ran `npm install` with, at least what I believe to be, relatively common dependencies (and not many of them), I have wasted an exorbitant amount of time just waiting for updates to install.  Additionally, because packages nest packages and it goes on ad infinitum, it quickly gets to a point where windows cannot even handle many of the file names.

**Comment by iarna** [comment]
My feeling is that anyone who would actually be inclined to use a `.npmignore-minimal` is already using `.npmignore` or `files` to limit what goes into their package. Indeed, I would encourage anyone putting together a module today to use `files` to explicitly select what they want included in their package.

I think the combination of `files` and something like what's described in [#6351](https://github.com/npm/npm/issues/6351) would meet the needs being expressed here. As such I'm going to close this issue.

**Comment by getify** [comment]
@iarna -- respectfully, that doesn't even remotely satisfy the use case(s) described here.

I am well aware that I can filter out files. That's not the point. The point is that some consumers of my package want a minimal set of files and other consumers want a lot more files. It is not acceptable to have to make separate packages for that purpose, nor is that scalable/sustainable from the naming perspective to have to add "-minimal" suffix to a package or something.

**Comment by othiym23** [comment]
I'll clarify the CLI team's position further:

We believe that the functionality described in this feature request isn’t worth baking into the CLI itself, because the arguments around what's "correct" and "useful" here are tedious and irresolvable. A significant fraction of users are going to be dissatisfied no matter what we do. Different people have different expectations around what should be included in packages, and npm offers a variety of tools to allow them to decide for themselves what belongs in their own published packages. A variety of other tools, some of them discussed in this thread, are available for others who want to slim down `node_modules` once the install is complete (potentially automatically via a lifecycle script). This is an excellent problem for third-party tools to address, because everybody can have the solution that works best for them.

The most compelling argument for baking something like this into the CLI is the IoT / small systems / constrained bandwidth use case. Unfortunately, it's not really adequate to only incorporate functionality that filters what's installed at runtime because of the power / bandwidth constraints implicit in the use case. To be an effective solution to the underlying problem would require publishing multiple versions of packages to the registry, or otherwise having some way of requesting only a minimal package at cache time. This would be an entirely new capability for the registry as well as the CLI, and it's not impossible that something like this would land as we work towards a new registry API, but this work is currently not on the registry roadmap (R&D is only just beginning on the design of the new API), and so is currently out of scope for the CLI.

**Comment by traviswimer** [comment]
I think @iarna made an important point:
> Indeed, I would encourage anyone putting together a module today to use files to explicitly select what they want included in their package.

Since I would imagine the "minimal" download is the more common use case, does NPM (or would they be willing to) officially encourage publishers to use `.npmignore` and `files` to ensure only the bare minimum number of files are included?

By this I mean the issue would be mentioned in the relevant documentation, such as: https://docs.npmjs.com/getting-started/publishing-npm-packages

My reasoning is that the issue @getify pointed out is caused by inconsistency across packages:
> The point is that some consumers of my package want a minimal set of files and other consumers want a lot more files.

Since most NPM packages are on GitHub and NPM supports install directly from GitHub, would it be possible to add a flag specifically for GitHub installs to cover the scenario of "other consumers want a lot more files"? This would get around the issue mentioned by @othiym23 that the "solution to the underlying problem would require publishing multiple versions of packages to the registry", since there are already additional versions published to GitHub.

**Comment by getify** [comment]
> I would imagine the "minimal" download is the more common use case

Not in my experience. I'd say probably 80% of the time I want everything because I'm a developer and I'm using npm to get dev source files. 20% of the time I want only the minimal files because I'm using npm for deploy.

And I'd say that not that many people have asked for fewer files from my packages, though some have. I also suspect that more people (but not a big majority) have wanted fewer files but have never asked because they assumed it wasn't possible.

What sucks here is that some people think one way should be the default, others think the other should be a default, but those in charge here are perfectly content to let only one of those two be handled directly, and the other path people just have to hack their way through.

I wish that this tool aspired to reasonably handle reasonable use-cases, even if that's *eventually*. Closing this issue says, "we're probably not gonna ever care about this." :/

> Since most NPM packages are on GitHub 

Not all are. This is my biggest frustration: when we just casually assume github is the panacea for limitations in the registry.

> there are already additional versions published to GitHub.

That is a bad solution. Am I supposed to maintain a separate branch on my repo where I, the author, manually cull out all but the minimal files, and then users are supposed to know that it's not just "go get it on github", but "go get it from this specific branch on github"?

These are all bad suggestions posed by people who seem from my reading to not understand or not care about the underlying issue.

**Comment by PixnBits** [comment]
Our build servers at work have no access to GitHub (or the Internet as a whole).

**Comment by traviswimer** [comment]
To clarify my previous comment:

- What I was suggesting is that if "minimal" packages were generally published to NPM, then it may be possible to achieve the "full" package use case through GIT repos.
- It would have been more reasonable if I had said "Since most NPM packages are in a public git repo" rather than GitHub specifically.

> That is a bad solution. Am I supposed to maintain a separate branch on my repo where I, the author, manually cull out all but the minimal files, and then users are supposed to know that it's not just "go get it on github", but "go get it from this specific branch on github"?

The scenario I suggested was basically "go get it on github". When I said "Additional versions" I meant in addition to what was released on NPM, not having multiple simultaneous version in a GIT repo. (minimal version on NPM, optional CLI flag to install full version from GIT repo)

> Our build servers at work have no access to GitHub (or the Internet as a whole).
I can see how a situation like this may be an issue, but if the build servers lack internet access, how can they install from NPM either?

To be clear, I'm certainly not arguing against the points made by @getify . But as he said NPM appears to be taking the position of "we're probably not gonna ever care about this." Although it would be great if NPM wanted to implement this feature, I think it is alternatively worth seeking a compromise, which was the intended purpose of my comment.

**Comment by PixnBits** [comment]
@traviswimer we have an internal npm registry of approved packages and internal packages

**Comment by DmitryEfimenko** [comment]
Huge +1.
We have about 60mb worth of **production** dependencies, but it's not the size that poses the problem, but the amount of files, which is around 10500. It becomes an issue when your deployment process zips the whole thing up and then unzips it at the production environment. Unzipping 10500 files is very very slow. If there was a way to only get a bare minimum (we don't care for tests, examples, documentation...), this would not be as big of an issue.

**Comment by ljharb** [comment]
That's a flaw in your deployment process, and unrelated to this problem. You should only be sending the files you need to production :-/

**Comment by getify** [comment]
@ljharb if this feature existed (and were used by the package author), fixing that flaw would be as trivial as adding --minimal to the CLI call. As it stands, they probably would have to do an incredibly manual/intricate process of figuring out only what they need and whitelist filtering out (maybe with grep or something) everything else, as it feeds filenes to the zip file process. Worse, if the package ever renames or moves any directory the their grep matching breaks. Your suggestion is, frankly, right in line with the other (moderate) condescension of this thread, which is to stay "go away you're doing it wrong".

What I can't understand is why people don't want to allow this option even if they would never use it? NPM would have more work, but users wouldn't. The default would be as it always has been, and only using --minimal would change anything. 

Shrugs.

**Comment by ljharb** [comment]
@getify I love this option - I want to publish all my files, but not impose burden on my consumers. The comment I was responding to isn't really related to that - naively zipping up `node_modules` doesn't necessarily exclude the things you don't use, even with the `--minimal` option. "doing it wrong" applies to that, not to wanting minimal installs.

**Comment by getify** [comment]
@ljharb- sorry if misinterpreted. i thought their implication was that 'minimal' would help with their zipping.

> doesn't necessarily exclude the things you don't use, even with the --minimal option

doesn't it tho? if all my node_modules are 'minimal' installs, why is it abjectly wrong to deploy by zipping it up? I've definitely done similar things in non-node land before.

**Comment by ljharb** [comment]
@getify Just because something is npm-installed doesn't mean you're actively using it in your application - for browser-side deps, that's why you'd use a bundler, and only send the bundle files. For server-side apps, it's not as important, and could also be resolved by walking the dependency graph, and only zipping up the files you need. I do agree it's something a `--minimal` option will help with, either way, but I want to stress that it's not about "there's lots of files sitting in node_modules" so much as "authors need a declarative way to indicate which files are needed in production, and which are not, without obliterating the latter from the package"

**Comment by getify** [comment]
@ljharb i think we're vehemently agreeing here. :)

**Comment by DmitryEfimenko** [comment]
@ljharb 
> You should only be sending the files you need to production
I even bolded the word "production" in my post because I knew this will come up.
It looks like rollup.js or webpack have capability to minify your code (including node_packages) and using these might take care of the whole issue completely.

**Comment by gbaumgart** [comment]
hi,
here is my current list to cleanup the mess most packages create on the hard-disc: 
```js
grunt.extendConfig({
    clean: {
        "platform": [
            dst + '/**/*.pdb',
            dst + '/**/*.o.d',
            dst + '/**/*.a.d',
            dst + '/**/*.obj',
            dst + '/**/*.lib',
            dst + '/**/*.exp',
            dst + '/**/MAKEFILE',
            dst + '/**/Makefile',
            dst + '/**/*.Makefile',
            dst + '/**/*.target.mk',
            dst + '/**/*.node.mk',
            dst + '/**/Cakefile',
            dst + '/**/Gruntfile.js',
            dst + '/**/Vagrantfile',
            dst + '/**/.tern-port',
            dst + '/**/*.cc',
            dst + '/**/*.c',
            dst + '/**/*.s', //asm
            dst + '/**/*.S', //asm
            dst + '/**/*.py',
            dst + '/**/*.pyc',
            dst + '/**/*.pl',
            dst + '/**/*.rb',
            dst + '/**/*.tlog',
            dst + '/**/*.sln',
            dst + '/**/.jshintrc',
            dst + '/**/.lint',
            dst + '/**/jsl.node.conf',
            dst + '/**/.eslintrc',
            dst + '/**/.editorconfig',
            dst + '/**/.jscs.json',
            dst + '/**/.npmignore',
            dst + '/**/.eslintignore',
            dst + '/**/*.dntrc',
            dst + '/**/*.cpp',
            dst + '/**/*.jpg',
            dst + '/**/*.png',
            dst + '/**/.gitmodules',
            dst + '/**/*.h',
            dst + '/**/*.patch',
            dst + '/**/_test',
            dst + '/**/LICENSE',
            dst + '/**/*LICENSE*',
            dst + '/**/*LICENCE*',
            dst + '/**/License',
            dst + '/**/license',
            //dst + '/**/*test.js*',    <-- some packages won't work with this
            //dst + '/**/*-test.js*',   <-- some packages won't work with this  
            dst + '/**/AUTHORS',
            dst + '/**/LICENSE.txt',
            dst + '/**/MIT-LICENSE.txt',
            dst + '/**/README',
            dst + '/**/README.md',
            dst + '/**/usage.txt',
            dst + '/**/*.md',
            dst + '/**/*.txt',
            dst + '/**/*.markdown',
            dst + '/**/Changelog',
            dst + '/**/CHANGES',
            dst + '/**/CHANGELOG',
            dst + '/**/.travis.yml',
            dst + '/**/appveyor.yml',
            dst + '/**/robot.html',
            dst + '/**/examples',
            dst + '/**/example',
            dst + '/**/example.js',
            dst + '/**/screenshots',
            dst + '/**/.gitattributes',
            dst + '/**/benchmarks',
            dst + '/**/*benchmark*',
            dst + '/**/coverage',
            dst + '/**/docs',
            dst + '/**/*.coffee',
            dst + '/**/tests',
            dst + '/**/*.vcxproj',
            dst + '/**/*.vcxproj.filters',
            dst + '/**/node-pre-gyp',
            dst + '/**/node-gyp',
            dst + '/**/gyp',
            dst + '/**/*.gypi'
        ]
    }
});
```
After running this, you can free around 80-90% space. I tried this against a few heavy apps of mine, all good.

- node-gyp is the biggest waster, accounting easily for hundreds of megabytes for debug symbols
- most people leave there tests stuff even in production mode, if you glob them out, you can get easily another 10% more

it would be great if node-js creates something as seen in electron, a virtual fs (one file) because if you profile on low-spec devices the file access is just insane. i know you could go over webpack but I never saw this working well with native modules.

g

**Comment by Mujaddadi** [comment]
+1. I have been wondering about it since the time I started using package managers, that why we get a whole lot of rubbish when we just want to use bootstrap.min.css or angular.min.js. I am sure most people do not modify angular.js before using in their apps. Beside that, it will reduce the unnecessary usage of bandwidth from both the user and from npm servers. Do we have any option now? since this issue is more than two year old.

**Comment by felixdorner** [comment]
+1.

Each package owner should be able to mark the minimum must-have files. This is totally optional.

NPM users then can easily decide for themselves if they want to load the full package or just the light version. This might work with a new flag in the CLI like --light. If there is no light version the CLI just shows an error.

Here are some benefits:

- Smaller node_modules folder takes up less disk space and less bandwidth (Yes I know - it's already pretty small)
- Cleaner node_modules folder (I like to keep my dev repo as clean as possible - there are others like me as well)
- Less load for NPM servers

I would love to see something like this on a future roadmap.

**Comment by getify** [comment]
Respectfully, +1'ing this nearly three year old thread is basically pointless.

The folks at npm have repeatedly stated in this thread that they don't think the use-case(s) here matter enough to warrant their attention. As a matter of fact, I'm still pretty sure they don't even understand the use case, based on the repeated non-solution answers given. I'd also bet that they've long since unsubscribed from this thread.

I don't know if they'd ever re-consider, but I doubt adding to **this** thread is going to bump their attention.

**Comment by felixdorner** [comment]
@getify maybe you should write a profound Medium article on improving NPM to get their attention. I will try Yarn next - maybe this is something for you: https://yarnpkg.com/en/docs/cli/clean
